### PR TITLE
feat: add mermaid support in codeblock

### PR DESCRIPTION
**Pull request by C1C0**
# Motivation

[MermaidJS](https://mermaid.live/edit#pako:eNqFkUFLAzEQhf_KkFMLBe-LCEWvRbF6kb2MyWQ37WYSJ9mClP5308a6tR7MKe99j0cys1c6GFKN8o6Nx9gygISQZ7NvYz4_WgDPlAhF91UBPDKQtaSz2zGldPsuN3fIBixhHoXSRW455uAxOw1aCnWBzxDgNU3Reu5PmR1BJt2z-xivA-ssmKkrdXFAZsfdb76UbvTEGcrr4x9q7FWdNf8ZVb2EMPyAJ2I4_jZiJDmbKxKPzlS57MNmqjG2T2GStt8kd6nNJm0HtVC-VpR97I-0VbknT61qytWgbFvV8qHksIx0_claNVlGWqgxmjKTB4edoD-bZFwOsqr7Pa15oSLyWwglYnFIdPgCF8SXkQ) is a user-friendly diagramming tool. It is known for making diagram creation easier and faster than writing detailed descriptions. By adding MermaidJS to Memos, I aim to enable users to include various types of diagrams, like mind maps, in their notes. This integration is intended to make the sharing of complex ideas more straightforward and visually appealing, helping to communicate information more effectively in memos.

Example:

image

image

### CODE REVIEW

File: web/src/components/MemoContent/CodeBlock.tsx [hunk 1]
```diff
@@ -28,24 +33,35 @@ const CodeBlock: React.FC<Props> = ({ language, content }: Props) => {
     // Skip error and use default highlighted code.
   }
 
+  if (formatedLanguage === MERMAID_LANGUAGE) {
+    languageSpecificClasses = "mermaid";
+  }
```
Comment (boojack) [inline web/src/components/MemoContent/CodeBlock.tsx:38]: Let's return the renderer directly as same as https://github.com/usememos/memos/pull/2971/files#diff-e7203928564407fe22b273bea681a744e690877dc88daf9cac949fc5f09e011cR23

File: web/src/components/MemoContent/MermaidBlock.tsx [hunk 2]
```diff
@@ -0,0 +1,18 @@
+import mermaid from "mermaid";
+import { useEffect } from "react";
+import { BaseProps } from "./types";
+
+interface Props extends BaseProps {
+  __html: string;
+}
+
+const CodeBlock: React.FC<Props> = ({ __html, className }: Props) => {
```
Comment (boojack) [inline web/src/components/MemoContent/MermaidBlock.tsx:9]:
```suggestion
const MermaidBlock: React.FC<Props> = ({ __html, className }: Props) => {
```

**Review by boojack — CHANGES_REQUESTED** [review]

File: web/src/components/MemoContent/CodeBlock.tsx (hunk 1 above)
Comment (C1C0) [reply web/src/components/MemoContent/CodeBlock.tsx]: @boojack Thank you for catching it up 👍 . Updated to return renderer directly.

Unfortunatelly, the link you've provided doesn't seem to point anywhere specific in the code.

File: web/src/components/MemoContent/MermaidBlock.tsx (hunk 2 above)
Comment (C1C0) [reply web/src/components/MemoContent/MermaidBlock.tsx]: 👍  Updated

**Comment by MehadND** [comment]
What about the pricing? Free?

File: web/src/components/MemoContent/MermaidBlock.tsx [hunk 3]
```diff
@@ -0,0 +1,25 @@
+import mermaid from "mermaid";
+import { useEffect, useRef } from "react";
+
+interface Props {
+  __html: string;
```
Comment (boojack) [inline web/src/components/MemoContent/MermaidBlock.tsx:5]:
```suggestion
  content: string;
```

File: web/src/components/MemoContent/CodeBlock.tsx [hunk 4]
```diff
@@ -19,6 +22,10 @@ const CodeBlock: React.FC<Props> = ({ language, content }: Props) => {
     return <div className="w-full overflow-auto !my-2" dangerouslySetInnerHTML={{ __html: content }} />;
   }
 
+  if (formatedLanguage === MERMAID_LANGUAGE) {
+    return <MermaidBlock __html={highlightedCode} />;
```
Comment (boojack) [inline web/src/components/MemoContent/CodeBlock.tsx:26]:
```suggestion
    return <MermaidBlock content={content} />;
```

**Review by boojack — APPROVED** [review]
LGTM, otherwise

**Pull request merged**